### ISSUE TITLE
Config not resolving dependencies when forcing class to bind from YAML.

### ISSUE THREAD

**Issue by rbudzko** [spring-projects/spring-boot#4034]
When binding YAML both properties and dependencies are resolved - see attached gist please. 

Unexpected behavior was encountered when class had been specified in YAML file - uncomment both commented lines in application.yml. Bean available to wire has nothing in common (or so I believe) with configuration resolved using Snake YAML. Judging by stacks two instances of Config are created - one by snake (value from application.yaml injected) and one as a @Configuration result (dependency auto wired + this instance is available in context for further dependency management).

I believe that this behavior is not expected - sorry to arrive without a patch, but I'd like to confirm desired behavior first.

Gist -> https://gist.github.com/rbudzko/f3641685cb34a7025a54

**Comment by wilkinsona** [comment]
I'm not sure that I've understood the problem. It sounds like you're expecting your application context to use the `Config` instance created by SnakeYAML, rather than one created by Spring?

The configuration that's loaded from YAML is used to configure existing beans, not to add beans to the application context. Assuming that I've understood your problem correctly, what you're trying to do will work for settings values on an `@ConfigurationProperties` bean but won't work as a means of creating the `@ConfigurationProperties` bean itself.

**Comment by rbudzko** [comment]
I do expect my application context use Config created by SnakeYAML, because when "!!org.rbudzko.Config" indicator is not used same instance of object has both property from YAML and Bean injected, so it is indeed part of application context (maybe I've miss assumed that it is created by SnakeYAML in this scenario).

- implementation is set by AutowiredAnnotationBeanPostProcessor on object A (single bean of config's type)
- primitive is set by PropertiesConfigurationFactory on object A (single bean of config's type)

When I explicitly force type in YAML things work differently:

- implementation is set byAutowiredAnnotationBeanPostProcessor  on object A (single bean of config's type)
- primitive is set by YamlProcessr ob object B

What confuses me is a fact that in first scenario  @Component + @ConfigurationProperties are combined into single bean with both fields configured but simple change (choosing explicit type to bind to) is resulting in @Component + @ConfigurationProperties being resolved as two totally different thing (two instances of config in JVM). 

It's a little inconsistent, isn't it?

**Comment by wilkinsona** [comment]
No, it's entirely consistent with the configuration you've provided. In the second scenario, you've told SnakeYAML to create a Config instance and you've told Spring Boot to create one. As a result you get two instances of Config. In the first scenario, SnakeYAML creates Lists and Maps of scalar values and passes them to Spring Boot so you only get a single Config instance.

Why are you forcing a type in your YAML? Spring Boot's configuration binding is smart enough to figure out the type for you, and the types of anything nested beneath your configuration properties bean for that matter.

**Comment by rbudzko** [comment]
I see! Now it makes sense to me (explanation about maps passed from snake to boot).

I'm forcing type, because in real scenario of mine I'd like to have list of same interface injected and this list's elements are implemented differently, so configured differently in YAML as well. In short - I mix simple implementation with configuration, because it's extremely convenient. When implementation became more complicated (required auto wiring other bean) I've encountered explained scenario, so I've refactored it.

I still would appreciate instance created by SnakeYAML becoming managed bean, but if current behavior is intended we should simply close this issue :-).
